blog/articles/distributed-tracing.md

```diff
@@ -0,0 +1,488 @@
+don’t have any ingress installed. No matter, a port-forward will suffice for our
+purposes: running `kubectl -n observability port-forward service/simplest-query
+16686` will give us our Jaeger dashboard will be accessible on
+[http://localhost:16686](http://localhost:16686).
+
+Next we’re going to create the [OpenTelemetry
+collector](https://opentelemetry.io/docs/collector/getting-started/), which will
+be responsible for receiving traces from our programs and forwarding them to
+Jaeger. The collector is an abstraction that lets us connect systems that speak
+different protocols to each other. Even if we only export Zipkin traces we can
+rely on the collector to translate them into the form that Jaeger can consume.
+[This collector
+definition](https://github.com/benmoss/knative-tracing/blob/main/otel.yaml) will
+tell the OpenTelemetry operator to create a collector that will listen for
+traces as if it’s a Zipkin instance, but export them both to logs for debugging
+as well as to our Jaeger instance.
+
+
+```
+kubectl apply -f - <<EOF
+apiVersion: opentelemetry.io/v1alpha1
+kind: OpenTelemetryCollector
+metadata:
+  name: otel
+  namespace: observability
+spec:
+  config: |
+    receivers:
+      zipkin:
+    exporters:
+      logging:
+      jaeger:
+        endpoint: "simplest-collector.observability:14250"
+        insecure: true
+
+    service:
+      pipelines:
+        traces:
+          receivers: [zipkin]
+          processors: []
+          exporters: [logging, jaeger]
+EOF
+```
+
+If all is well we should now see 3 pods are running in the `observability`
+namespace: our Jaeger operator, our Jaeger instance, and the OpenTelemetry
+collector.
+
+Finally we can configure Eventing and Serving to point all of their traces to
+our collector:
+
+
+```
+for ns in knative-eventing knative-serving; do
+  kubectl patch --namespace "$ns" configmap/config-tracing \
+   --type merge \
+   --patch '{"data":{"backend":"zipkin","zipkin-endpoint":"http://otel-collector.observability:9411/api/v2/spans", "debug": "true"}}'
+done
+```
+
+
+The `debug` flag here tells Knative to send _all_ traces to our collector,
+whereas in a real-world deployment you’d probably want to set a sample rate to
+only get a representative subset of your traces.
+
+
+## Hello, world?
+
+Now that our tracing infrastructure is all deployed and configured, we can start
+taking advantage of it by deploying some services. We can deploy [the heartbeat
+image](https://github.com/knative/eventing/blob/41e36eea88aada0f73e98ff5fb39b6509135dda1/cmd/heartbeats/main.go)
+as a
+[ContainerSource](https://knative.dev/docs/eventing/samples/container-source/)
+to test and see that everything is wired up correctly:
+
+
+```
+kubectl apply -f - <<EOF
+apiVersion: sources.knative.dev/v1
+kind: ContainerSource
+metadata:
+  name: heartbeats
+spec:
+  template:
+    spec:
+      containers:
+        - image: benmoss/heartbeats:latest
```

Comment:
Can we use the heartbeats upstream image instead of the personal image?
```suggestion
        - image: gcr.io/knative-releases/knative.dev/eventing/cmd/heartbeats
```

Reply:
The problem I ran into with the upstreams was that they didn't have tracing support yet, https://github.com/knative/eventing/pull/5631 hasn't landed in the official images yet afaik

Reply:
We can wait for that support to land, but I guess it won't be until 0.26